tmt/utils.py

```diff
@@ -2544,3 +2553,80 @@ def find_fmf_root(path: str) -> List[str]:
         raise MetadataError(f"No fmf root present inside '{path}'.")
     fmf_roots.sort(key=lambda path: len(path))
     return fmf_roots
+
+
+# A type for callbacks given to wait()
+WaitCheckType = Callable[[], T]
+
+
+def wait(
+    parent: Common,
+    check: WaitCheckType[T],
+    timeout: datetime.timedelta,
+    tick: float = 30.0,
+    tick_increase: float = 1.0
+) -> T:
+    """
+    Wait for a condition to become true.
+
+    To test the condition state, a ``check`` callback is called every ``tick``
+    seconds until ``check`` reports a success. The callback may:
+
+    * decide the condition has been fulfilled. This is a successfull outcome,
+      ``check`` shall then simply return, and waiting ends. Or,
+    * decide more time is needed. This is not a successfull outcome, ``check``
+      shall then raise :py:clas:`WaitingIncomplete` exception, and ``wait()``
+      will try again later.
+
+    :param parent: "owner" of the wait process. Used for its logging capability.
+    :param check: a callable responsible for testing the condition. Accepts no
+        arguments. To indicate more time and attempts are needed, the callable
+        shall raise :py:class:`WaitingIncomplete`, otherwise it shall return
+        without exception. Its return value will be propagated by ``wait()`` up
+        to ``wait()``'s. All other exceptions raised by ``check`` will propagate
+        to ``wait()``'s caller as well, terminating the wait.
+    :param timeout: amount of time ``wait()`` is alowed to spend waiting for
+        successfull outcome of ``check`` call.
+    :param tick: how many seconds to wait between two consecutive calls of
+        ``check``.
+    :param tick_increase: a multiplier applied to ``tick`` after every attempt.
+    :returns: value returned by ``check`` reporting success.
+    :raises GeneralError: when ``tick`` is not a positive integer.
+    :raises WaitingTimedOutError: when time quota has been consumed.
+    """
+
+    if tick <= 0:
+        raise GeneralError('Tick must be a positive integer')
+
+    deadline = datetime.datetime.utcnow() + timeout
+
+    parent.debug(
+        'wait',
+        f"waiting for condition '{check.__name__}' with timeout {timeout},"
+        f"deadline {deadline}, check every {tick} seconds")
+
+    while True:
+        now = datetime.datetime.utcnow()
+
+        if now > deadline:
+            raise WaitingTimedOutError()
+
+        try:
+            ret = check()
```

Comment:
Before I forget about this, we should check here if `check()` didn't actually take longer than `deadline`. This can and does happen, e.g. with #1293 , if we use `tmt-reboot -t 1`, `ssh` could take multiple seconds which makes it exceed timeout (which should result in an error)

Reply:
That can very well happen. What you would like to see is an extra check after *a successful* `check()` call, making sure that even a successful call would result in a timeout error if it took longer than allowed, correct? That can be arranged.

Reply:
Yes, I think that would make the most sense since it would make the helper 100% respect the deadline which may be useful if we want to be strict with stuff such as timeouts (for example the reboot timeout). And I don't see a better way how to achieve this than comparing time even after `check()`

Reply:
I see no better way either. Noted.

Reply:
Fixed after rebasing, there should be an extra check to make sure even successful `check()` call turns into a failure if it took too long.